pkg/api-server/resource_endpoints.go

```diff
@@ -532,17 +532,45 @@ func (r *resourceEndpoints) validateLabels(rm model.ResourceMeta) validators.Val
 		}
 	}
 
-	for _, k := range maps.SortedKeys(rm.GetLabels()) {
+	if r.mode != config_core.Global {
+		zoneTag, ok := resource.GetMeta().GetLabels()[mesh_proto.ZoneTag]
+		if ok && zoneTag != r.zoneName {
+			err.AddViolationAt(validators.Root().Key(mesh_proto.ZoneTag), fmt.Sprintf("%s label should have %s value", mesh_proto.ZoneTag, r.zoneName))
+		}
+	}
+
+	if r.descriptor.IsPluginOriginated && r.descriptor.IsPolicy {
+		r.validatePolicyRole(resource)
+	}
+
+	for _, k := range maps.SortedKeys(resource.GetMeta().GetLabels()) {
 		for _, msg := range validation.IsQualifiedName(k) {
 			err.AddViolationAt(validators.Root().Key(k), msg)
 		}
-		for _, msg := range validation.IsValidLabelValue(rm.GetLabels()[k]) {
+		for _, msg := range validation.IsValidLabelValue(resource.GetMeta().GetLabels()[k]) {
 			err.AddViolationAt(validators.Root().Key(k), msg)
 		}
 	}
 	return err
 }
 
+func (r *resourceEndpoints) validatePolicyRole(resource rest.Resource) validators.ValidationError {
+	var err validators.ValidationError
+	policyObject, ok := resource.GetSpec().(core_model.Policy)
+	if !ok {
+		return err
+	}
+	policyRole, ok := resource.GetMeta().GetLabels()[mesh_proto.PolicyRoleLabel]
+	if !ok {
+		return err
```

Comment:
same, you're returning an empty error, just use [PolicyRole](https://github.com/kumahq/kuma/blob/d89e3e53960da72f261229f699b22c3e2d3273e8/pkg/core/resources/model/resource.go#L493) it always returns a value

Reply:
I am returning empty error since missing policy role is not a problem here. I cannot use `PolicyRole` method here, as this the resource types are not compatible. This is a `rest.Resource`

Reply:
Just replace `PolicyRole` method with:
```go
func PolicyRole(rm ResourceMeta) mesh_proto.PolicyRole {
	if rm.GetLabels() == nil || rm.GetLabels()[mesh_proto.PolicyRoleLabel] == "" {
		return mesh_proto.SystemPolicyRole
	}
	return mesh_proto.PolicyRole(rm.GetLabels()[mesh_proto.PolicyRoleLabel])
}
```
it doesn't actually need the whole Resource

Reply:
changed